crates/provider/src/helpers.rs

```diff
@@ -0,0 +1,180 @@
+use std::{error::Error, fs::File};
+
+use crate::{native::FileSystem, shared::types::LocalFile};
+
+#[derive(Debug, PartialEq)]
+pub enum Operation {
+    // DeleteFile { path: String },
+    // DeleteDir,
+    // LinkFile,
+    Copy { from: String, to: String },
+    ReadFile { path: String },
+    CreateFile { path: String },
+    CreateDir { path: String },
+    OpenFile { path: String },
+    Write { path: String, content: String },
+}
+
+#[derive(Debug)]
+pub struct MockFilesystem {
+    copy_error:        Option<Box<dyn Error + Send + Sync>>,
+    create_dir_error:  Option<Box<dyn Error + Send + Sync>>,
+    create_file_error: Option<Box<dyn Error + Send + Sync>>,
+    open_file_error:   Option<Box<dyn Error + Send + Sync>>,
+    read_file_error:   Option<Box<dyn Error + Send + Sync>>,
+    write_error:       Option<Box<dyn Error + Send + Sync>>,
+    pub operations:    Vec<Operation>,
+}
+
+impl MockFilesystem {
+    pub fn new() -> Self {
+        Self {
+            copy_error:        None,
+            create_dir_error:  None,
+            open_file_error:   None,
+            create_file_error: None,
+            read_file_error:   None,
+            write_error:       None,
+            operations:        vec![],
+        }
+    }
+
+    fn with_create_dir_error(error: impl Error + Send + Sync + 'static) -> Self {
```

Comment:
This argument type can be extracted to a generic at the impl level to make the code cleaner:

```rust
impl<E> MockFilesystem
where Err: Error + send + Sync + 'static 
 {
...
}
```

and then be used in all following methods like this:

```suggestion
    fn with_create_dir_error(error: E) -> Self {
```